### ISSUE TITLE
perf: training cost scales super-linearly with sequence length — loss_and_grad dominance

### ISSUE THREAD

**Issue by teilomillet** [teilomillet/textpolicy#55]
## Context

PR #53 profiling revealed that at paper-realistic scales, **training shifts from a minor cost to nearly half of total time**, and the balance between rollout and training changes dramatically:

| Config | rollout_collect | train | rollout % | train % |
|--------|----------------|-------|-----------|---------|
| G=8, 256tok | 48.53s | 10.00s | **83%** | 17% |
| G=16, 512tok | 214.61s | 178.64s | **55%** | **45%** |

Going from 256→512 tokens and 8→16 episodes (4x more tokens), training cost increased **18x** (9.8s → 178s), not 4x. This is consistent with attention's O(n²) scaling on longer sequences.

Within training, `loss_and_grad` is 99.8% of the phase — the compiled forward+backward pass dominates everything.

Reference papers for realistic scales:
- GTPO ([arXiv:2508.04349](https://arxiv.org/abs/2508.04349)): G=16, 4096 max response, lr=1e-6
- HICRA ([arXiv:2509.03646](https://arxiv.org/abs/2509.03646)): 16K–32K context length

### 1. `mx.compile` for training (already available)
- `--compile-training true` wraps `loss_and_grad_fn` with `mx.compile`
- Should fuse operations and reduce kernel launch overhead
- Currently defaults to `false` for HICRA compatibility — worth investigating if the compatibility issue is resolved

### 2. Gradient checkpointing
- Trade compute for memory: recompute activations during backward instead of storing them
- Especially valuable at 512+ token sequences where activation memory dominates
- MLX supports `mx.checkpoint` for selective recomputation

### 3. Prompt/response masking in loss
- Only response tokens contribute to the loss (prompt tokens are masked)
- If the forward pass still processes prompt tokens through all layers, masking earlier could save compute
- Current implementation in `_loss_fn` (trainer.py:403) processes the full concatenated sequence

### 4. Sequence packing / padding efficiency
- `_pack_episodes` right-pads to max sequence length in the batch
- If episode lengths vary significantly, shorter episodes waste compute on padding
- Sequence packing (multiple episodes per row) could improve utilization

### 5. LoRA rank scaling
- Currently using rank=2 (minimal). Higher ranks increase training cost linearly
- At rank=2, the LoRA overhead is negligible vs. base model forward pass
- But if rank is increased for quality, training cost will grow

## How to reproduce

```bash
# Compare scaling behavior:
uv run python experiments/countdown_reasoning_lora.py \
  --steps 3 --episodes-per-step 8 --batch-size 8 \
  --max-tokens 256 --profile-training --output results/profile_256tok

uv run python experiments/countdown_reasoning_lora.py \
  --steps 3 --episodes-per-step 16 --batch-size 8 \
  --max-tokens 512 --profile-training --output results/profile_512tok
```

Compare "Trainer-internal split" in both Amdahl summaries.

## Labels

Exploratory — documents the scaling behavior for future optimization work. The profiling infrastructure (PR #53) makes impact measurement straightforward.
